### PR TITLE
[Discussion] Rewrite of the Route Drawer

### PR DESCRIPTION
**Pull request by RoiEXLab**
I did some experiments with the apache commons math library in order to get a better route drawing system running.
Since the implementation of a new library needs a discussion well here you go...
screenshot
That's what I got so far...
Tell me what yout think...
(The way the line is drawn has to be changed...)

---
This change is [Reviewable](https://reviewable.io/reviews/triplea-game/triplea/654)

### CODE REVIEW

File: src/games/strategy/triplea/ui/MapRouteDrawer.java [hunk 1]
```diff
@@ -6,295 +6,162 @@
...
+    final int jointsize = 10;
+    
+    if(route.getAllTerritories().size() <= 1){
+      drawLineWithTranslate(graphics, new Line2D.Float(routeDescription.getStart(), routeDescription.getEnd()),
+        view.getXOffset(), view.getYOffset());
+      graphics.fillOval((routeDescription.getEnd().x - xOffset) - jointsize / 2, (routeDescription.getEnd().y - yOffset) - jointsize / 2, jointsize, jointsize);
+    }
+    else{
+      final double[] index = getIndex(points);
+      final float stepSize = 0.01f;//TODO calculating a step size that makes sense
```
Comment (RoiEXLab) [inline src/games/strategy/triplea/ui/MapRouteDrawer.java]: This step size needs to be calculated to be a pixel...

File: src/games/strategy/triplea/ui/MapRouteDrawer.java [hunk 2]
```diff
@@ -6,295 +6,162 @@
...
+      PolynomialSplineFunction xcurve = new SplineInterpolator().interpolate(index, pointsXToDoubleArray(points));
+      PolynomialSplineFunction ycurve = new SplineInterpolator().interpolate(index, pointsYToDoubleArray(points));
+      double[] xcoords = getCoords(xcurve, stepSize);
+      double[] ycoords = getCoords(ycurve, stepSize);
+      for(int i = 1; i < xcoords.length; i++){
+        //TODO maybe a line is not the best way to draw this...
+        drawLineWithTranslate(graphics, new Line2D.Double(xcoords[i-1], ycoords[i-1], xcoords[i], ycoords[i]), view.getXOffset(), view.getYOffset());
```
Comment (RoiEXLab) [inline src/games/strategy/triplea/ui/MapRouteDrawer.java]: If the stepSize is calculated to be a pixel we can draw pixel by pixel

**Comment by frigoref** [comment]
Review status: 0 of 2 files reviewed at latest revision, 12 unresolved discussions.

---

*[src/games/strategy/triplea/ui/MapRouteDrawer.java, line 35 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/654#-KH9WcItJHU9qJ5BiMPV:-KH9WcIuK-K7KCgcDlDa:-2004900413) ([raw file](https://github.com/triplea-game/triplea/blob/fa9c54a1ffa662dd314f94d48072021580a18312/src/games/strategy/triplea/ui/MapRouteDrawer.java#L35)):*
Would recommend moving this down to the place where the graphics variable is used the first time

---

*[src/games/strategy/triplea/ui/MapRouteDrawer.java, line 39 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/654#-KH9Ww7-NsBtyzwjdjMI:-KH9Ww7-NsBtyzwjdjMJ:-965668888) ([raw file](https://github.com/triplea-game/triplea/blob/fa9c54a1ffa662dd314f94d48072021580a18312/src/games/strategy/triplea/ui/MapRouteDrawer.java#L39)):*
Maybe rename to getRoutePoints. In general the code is hard to understand due to short 

---

*[src/games/strategy/triplea/ui/MapRouteDrawer.java, line 45 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/654#-KH9aOrhB0jwQJ1AkHMt:-KH9aOrhB0jwQJ1AkHMu:-1530876928) ([raw file](https://github.com/triplea-game/triplea/blob/fa9c54a1ffa662dd314f94d48072021580a18312/src/games/strategy/triplea/ui/MapRouteDrawer.java#L45)):*
According to your code in getPoints() those could be null, shouldn't that be handled?
Also, maybe it makes sense to get local references routeStartPoint and routeEndPoint instead of getting it always from the routeDescription variable.

---

*[src/games/strategy/triplea/ui/MapRouteDrawer.java, line 52 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/654#-KH9v6XGu201xy4U34zr:-KH9v6XGu201xy4U34zs:-1472814659) ([raw file](https://github.com/triplea-game/triplea/blob/fa9c54a1ffa662dd314f94d48072021580a18312/src/games/strategy/triplea/ui/MapRouteDrawer.java#L52)):*
Still getting errors:
Exception: number of points (2)
org.apache.commons.math3.analysis.interpolation.SplineInterpolator.interpolate(SplineInterpolator.java:74)
games.strategy.triplea.ui.MapRouteDrawer.drawRoute(MapRouteDrawer.java:52)

---

*[src/games/strategy/triplea/ui/MapRouteDrawer.java, line 53 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/654#-KH9_8yljyO5cy8ccfqC:-KH9_8yljyO5cy8ccfqD:1856629500) ([raw file](https://github.com/triplea-game/triplea/blob/fa9c54a1ffa662dd314f94d48072021580a18312/src/games/strategy/triplea/ui/MapRouteDrawer.java#L53)):*
Use same SplintInterpolator object

---

*[src/games/strategy/triplea/ui/MapRouteDrawer.java, line 60 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/654#-KH9_ngb_2C0F788LzXI:-KH9_ngc0zev2c7arqGq:1375610433) ([raw file](https://github.com/triplea-game/triplea/blob/fa9c54a1ffa662dd314f94d48072021580a18312/src/games/strategy/triplea/ui/MapRouteDrawer.java#L60)):*
The whole else case of route.getAllTerritories().size() <= 1 could be moved to a separate method.

---

*[src/games/strategy/triplea/ui/MapRouteDrawer.java, line 79 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/654#-KH9oRPL2KXOvOGN0vmA:-KH9oRPM6AMwxG3zQ6Fd:-445772412) ([raw file](https://github.com/triplea-game/triplea/blob/fa9c54a1ffa662dd314f94d48072021580a18312/src/games/strategy/triplea/ui/MapRouteDrawer.java#L79)):*
Try index[0] = 0 and index[i+1] = index[i] + sqrtDistance(point[i],point[i+1]).
It is going to be more accurate as it is taking into account the distance between the points.

---

*[src/games/strategy/triplea/ui/MapRouteDrawer.java, line 95 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/654#-KH9W-ddqJCpBqMBFjab:-KH9W-ddqJCpBqMBFjac:599685851) ([raw file](https://github.com/triplea-game/triplea/blob/fa9c54a1ffa662dd314f94d48072021580a18312/src/games/strategy/triplea/ui/MapRouteDrawer.java#L95)):*
Better to take a local List<Territory> variable for the territories at the beginning of the method instead of calling it collecting everytime.

---

*[src/games/strategy/triplea/ui/MapRouteDrawer.java, line 113 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/654#-KH9tFyNtN0lab1KKHTb:-KH9tFyNtN0lab1KKHTc:2094185779) ([raw file](https://github.com/triplea-game/triplea/blob/fa9c54a1ffa662dd314f94d48072021580a18312/src/games/strategy/triplea/ui/MapRouteDrawer.java#L113)):*
Instead of looping again maybe both should be returned after one loop. Something like splitToDoubleArrays

---

*[src/games/strategy/triplea/ui/MapRouteDrawer.java, line 179 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/654#-KH9uS0y9WMysorZO--T:-KH9uS0y9WMysorZO--U:-772285568) ([raw file](https://github.com/triplea-game/triplea/blob/daac4495cb151d1eb82de6c3f8a7e2ed13109476/src/games/strategy/triplea/ui/MapRouteDrawer.java#L179)):*
Why did this scrollWrap-part for cursorImage got removed?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/triplea-game/triplea/654)*

**Comment by RoiEXLab** [comment]
Review status: 0 of 2 files reviewed at latest revision, 12 unresolved discussions.

---

*[src/games/strategy/triplea/ui/MapRouteDrawer.java, line 45 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/654#-KH9aOrhB0jwQJ1AkHMt:-KH9wxQsUEVBaIQC28WF:-1527519438) ([raw file](https://github.com/triplea-game/triplea/blob/fa9c54a1ffa662dd314f94d48072021580a18312/src/games/strategy/triplea/ui/MapRouteDrawer.java#L45)):*
Good Point

---

*[src/games/strategy/triplea/ui/MapRouteDrawer.java, line 52 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/654#-KH9v6XGu201xy4U34zr:-KH9wI3E-WrbuhamXtMk:-640472748) ([raw file](https://github.com/triplea-game/triplea/blob/fa9c54a1ffa662dd314f94d48072021580a18312/src/games/strategy/triplea/ui/MapRouteDrawer.java#L52)):*
On which Map did you test this? This error only occurs, if there are less then 3 points...

---

*[src/games/strategy/triplea/ui/MapRouteDrawer.java, line 79 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/654#-KH9oRPL2KXOvOGN0vmA:-KH9xL0j5bTXnhQ0v0Q1:-1215757319) ([raw file](https://github.com/triplea-game/triplea/blob/fa9c54a1ffa662dd314f94d48072021580a18312/src/games/strategy/triplea/ui/MapRouteDrawer.java#L79)):*
Yeah this makes sense I guess

---

*[src/games/strategy/triplea/ui/MapRouteDrawer.java, line 179 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/654#-KH9uS0y9WMysorZO--T:-KH9xUUq7rtoeYvA-8s5:-588807044) ([raw file](https://github.com/triplea-game/triplea/blob/daac4495cb151d1eb82de6c3f8a7e2ed13109476/src/games/strategy/triplea/ui/MapRouteDrawer.java#L179)):*
Because I think it's better to have the Cursor on Top of the screen - there were you cursor normally would be... If your cursor is centered, there is no need to wrap it around...

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/triplea-game/triplea/654)*

**Comment by frigoref** [comment]
Review status: 0 of 2 files reviewed at latest revision, 12 unresolved discussions.

---

*[src/games/strategy/triplea/ui/MapRouteDrawer.java, line 52 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/654#-KH9v6XGu201xy4U34zr:-KH9yPsuU69BuL8Q-PCY:-79110481) ([raw file](https://github.com/triplea-game/triplea/blob/fa9c54a1ffa662dd314f94d48072021580a18312/src/games/strategy/triplea/ui/MapRouteDrawer.java#L52)):*
Map: 270BC

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/triplea-game/triplea/654)*

**Comment by RoiEXLab** [comment]
Review status: 0 of 2 files reviewed at latest revision, 12 unresolved discussions.

---

*[src/games/strategy/triplea/ui/MapRouteDrawer.java, line 79 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/654#-KH9oRPL2KXOvOGN0vmA:-KHBC8jHk4xhuULbln2Q:-1037391017) ([raw file](https://github.com/triplea-game/triplea/blob/fa9c54a1ffa662dd314f94d48072021580a18312/src/games/strategy/triplea/ui/MapRouteDrawer.java#L79)):*
For whatever reason this doesn't work...

---

*[src/games/strategy/triplea/ui/MapRouteDrawer.java, line 113 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/654#-KH9tFyNtN0lab1KKHTb:-KHB6INrjkvkMau6XKMl:-63036739) ([raw file](https://github.com/triplea-game/triplea/blob/fa9c54a1ffa662dd314f94d48072021580a18312/src/games/strategy/triplea/ui/MapRouteDrawer.java#L113)):*
How would you return 2 arrays at the same time? I don't think that a Map should be used here...

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/triplea-game/triplea/654)*

**Comment by RoiEXLab** [comment]
Review status: 0 of 2 files reviewed at latest revision, 12 unresolved discussions.

---

*[src/games/strategy/triplea/ui/MapRouteDrawer.java, line 52 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/654#-KH9v6XGu201xy4U34zr:-KHBKeo2RocRkJ7F3-uT:-497468049) ([raw file](https://github.com/triplea-game/triplea/blob/fa9c54a1ffa662dd314f94d48072021580a18312/src/games/strategy/triplea/ui/MapRouteDrawer.java#L52)):*
Fixed

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/triplea-game/triplea/654)*

**Comment by frigoref** [comment]
Reviewed 1 of 2 files at r1.
Review status: 1 of 2 files reviewed at latest revision, 11 unresolved discussions.

---

*[src/games/strategy/triplea/ui/MapRouteDrawer.java, line 79 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/654#-KH9oRPL2KXOvOGN0vmA:-KHBWOFFOh3Z9bnSc6O9:-2042308518) ([raw file](https://github.com/triplea-game/triplea/blob/fa9c54a1ffa662dd314f94d48072021580a18312/src/games/strategy/triplea/ui/MapRouteDrawer.java#L79)):*
What is not working?

---

*[src/games/strategy/triplea/ui/MapRouteDrawer.java, line 113 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/654#-KH9tFyNtN0lab1KKHTb:-KHBVyK0H8U1W1Xzqknr:-761990588) ([raw file](https://github.com/triplea-game/triplea/blob/fa9c54a1ffa662dd314f94d48072021580a18312/src/games/strategy/triplea/ui/MapRouteDrawer.java#L113)):*
If you pass them to the method splitToDoubleArrays(final Point[] points, double[] xValues, double[] yValues)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/triplea-game/triplea/654)*

**Comment by RoiEXLab** [comment]
Review status: 1 of 2 files reviewed at latest revision, 10 unresolved discussions.

---

*[src/games/strategy/triplea/ui/MapRouteDrawer.java, line 79 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/654#-KH9oRPL2KXOvOGN0vmA:-KHC2osfd3xrXrNLmm-6:-380603488) ([raw file](https://github.com/triplea-game/triplea/blob/fa9c54a1ffa662dd314f94d48072021580a18312/src/games/strategy/triplea/ui/MapRouteDrawer.java#L79)):*
When changin the index calculation like you said, the route gets drawn to about 5% the rest of the route is not visible

---

*[src/games/strategy/triplea/ui/MapRouteDrawer.java, line 113 \[r1\]](https://reviewable.io:443/reviews/triplea-game/triplea/654#-KH9tFyNtN0lab1KKHTb:-KHCMHjuGV0OSr5Z8UIq:-141136761) ([raw file](https://github.com/triplea-game/triplea/blob/fa9c54a1ffa662dd314f94d48072021580a18312/src/games/strategy/triplea/ui/MapRouteDrawer.java#L113)):*
I still don't get it D: I'm probably too tired...

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/triplea-game/triplea/654)*

**Comment by RoiEXLab** [comment]
@frigoref Tommotow is the last chance to tell me about something that needs to be changed... I'll be in London until Friday - without my PC

**Comment by frigoref** [comment]
Here is my suggestion which I roughly tested, so it might need some more rework.
Note that the stepSize should not be too small in order to avoid loosing some of the GeneralPath smoothing via rounding issues when converting double curve values to int pixel values.
```

  private static final double maxStepSizeCurveXValue = 25f;

  /**
   * Interpolates x and y values of points to generate a smooth path which includes the original points.
   *
   * @param points - points array
   * @return smooth path through provided points.
   */
  private static GeneralPath getSmoothPath(final Point[] points) {
    final GeneralPath result = new GeneralPath();
    // Interpolates x and y values
    final double[] curveXValues = getIndex(points);

    final Tuple<PolynomialSplineFunction, PolynomialSplineFunction> curves =
        buildSplineFunctions(points, curveXValues);
    final PolynomialSplineFunction xCurve = curves.getFirst();
    final PolynomialSplineFunction yCurve = curves.getSecond();
    // Build smooth path from interpolated curves including original points
    double curveXvalue = 0;
    result.moveTo(xCurve.value(curveXvalue), yCurve.value(curveXvalue));
    for (int curveXValueIndex = 1; curveXValueIndex < curveXValues.length; curveXValueIndex++) {
      // Get even currentStepSize between current and previous curveXValues
      final double diffCurveXValues = curveXValues[curveXValueIndex] - curveXValues[curveXValueIndex - 1];
      final int stepsUntilNextCurveXValue = (int) Math.ceil((diffCurveXValues) / maxStepSizeCurveXValue);
      final double currentStepSize = diffCurveXValues / stepsUntilNextCurveXValue;
      for (curveXvalue += currentStepSize; curveXvalue < curveXValues[curveXValueIndex]; curveXvalue +=
          currentStepSize) {
        result.lineTo(xCurve.value(curveXvalue), yCurve.value(curveXvalue));
      }
      curveXvalue = curveXValues[curveXValueIndex];
      result.lineTo(xCurve.value(curveXvalue), yCurve.value(curveXvalue));
    }
    return result;
  }

  private static Tuple<PolynomialSplineFunction, PolynomialSplineFunction> buildSplineFunctions(final Point[] points,
      final double[] curveXValues) {
    final double[] xCurveYValues = new double[points.length];
    final double[] yCurveYValues = new double[points.length];
    for (int i = 0; i < points.length; i++) {
      xCurveYValues[i] = points[i].getX();
      yCurveYValues[i] = points[i].getY();
    }
    return Tuple.of(splineInterpolator.interpolate(curveXValues, xCurveYValues),
        splineInterpolator.interpolate(curveXValues, yCurveYValues));
  }

```
Call:
`graphics.draw(getSmoothPath(points));`

**Comment by RoiEXLab** [comment]
Well close enough @frigoref 
screenshot 11
screenshot 12
You forgot to subtract the xOffset and yOffset value provided by `view.getYOffset();` and `view.getXOffset();`

I tested this with both versions of the get Index method...

**Comment by frigoref** [comment]
What were you painting there? Shouldn't we have smooth lines?
Have you thought about the issue when the path line crosses a border to another territory which does not belong to the route? It is buggy me since yesterday night ...

**Comment by RoiEXLab** [comment]
I copy and pasted your code into my branch... the screenshots were the result...
Maybe I did something wrong? I updated my [Gist](https://gist.github.com/RoiEXLab/5907b89487b3a465e4b2ca4b4a207859) to  match the code I currently habe setup...

**Comment by RoiEXLab** [comment]
And that Issue with the path being too long etc...
If you mean, that the path is just too - well inaccurate - the we could probably fix this by adding knots to out point array in order to shorten those paths...
This would be a little bit more complicated though.
If you mean something else, please explain what you mean...

**Comment by RoiEXLab** [comment]
btw The idea of storing our path into a geometric path object is very good...

**Comment by frigoref** [comment]
1. Screenshot showed wrong result as the path needs to be transformed with the view offsets:
```
      final GeneralPath path = getSmoothPath(points);
      path.transform(new AffineTransform(1, 0, 0, 1, -xOffset, -yOffset));
      graphics.draw(path);
```

2. Smoothing didn't work due to lineTo() call on path. I optimized it as follows:
```
  private static final double minStepSizeCurveXValue = 5f;
  private static final int maxPointsNeededForSmoothing = 3;

  /**
   * Generates a smooth path which includes the original points.
   *
   * @param points - points array
   * @return smooth path through provided points.
   */
  private static GeneralPath getSmoothPath(final Point[] points) {
    final GeneralPath result = new GeneralPath();
    if (points.length > 1) {
      result.moveTo(points[0].x, points[0].y);
    }
    if (points.length == 2) {
      result.lineTo(points[1].x, points[1].y);
    } else {
      addSmoothSegmentsToPath(result, points);
    }
    return result;
  }

  /**
   * Interpolates x and y values of the points-path and adds smooth segments from point to point to the path using the
   * interpolation result.
   *
   * @param path - path to be enhanced by smooth segments
   * @param points - original points
   */
  private static void addSmoothSegmentsToPath(final GeneralPath path, final Point[] points) {
    // Interpolates x and y values in separate curves
    final double[] curvesXValues = getIndex(points);
    final Tuple<PolynomialSplineFunction, PolynomialSplineFunction> curves =
        buildSplineFunctions(points, curvesXValues);
    final PolynomialSplineFunction xCurve = curves.getFirst();
    final PolynomialSplineFunction yCurve = curves.getSecond();
    // For each given point (to which one specific x value in curveXValues belongs)
    // use spline functions to get additional points for building the smooth path.
    // But no more needed than maxPointsNeededForDrawing.
    for (int curvesXValueIndex = 1; curvesXValueIndex < curvesXValues.length; curvesXValueIndex++) {
      // Get even currentStepSize between current and previous curvesXValues
      final double diffCurvesXValues = curvesXValues[curvesXValueIndex] - curvesXValues[curvesXValueIndex - 1];
      final int stepsUntilNextCurveXValue =
          Math.min(maxPointsNeededForSmoothing, (int) Math.ceil((diffCurvesXValues) / minStepSizeCurveXValue));
      final double currentStepSize = diffCurvesXValues / stepsUntilNextCurveXValue;
      // Collect interpolated values from xCurve and yCurve after each step
      final double xCurveValues[] = new double[stepsUntilNextCurveXValue];
      final double yCurveValues[] = new double[stepsUntilNextCurveXValue];
      // intermediate points from curve
      int curvesValueIndex = 0;
      for (double curvesXValue = curvesXValues[curvesXValueIndex - 1]; curvesValueIndex < stepsUntilNextCurveXValue
          - 1; ++curvesValueIndex) {
        curvesXValue += currentStepSize;
        xCurveValues[curvesValueIndex] = xCurve.value(curvesXValue);
        yCurveValues[curvesValueIndex] = yCurve.value(curvesXValue);
      }
      // next given point
      xCurveValues[curvesValueIndex] = points[curvesXValueIndex].x;
      yCurveValues[curvesValueIndex] = points[curvesXValueIndex].y;
      addSegmentToPath(path, xCurveValues, yCurveValues);
    }
  }

  private static void addSegmentToPath(final GeneralPath path, final double[] xCurveValues,
      final double[] yCurveValues) {
    switch (xCurveValues.length) {
      case 3:
        path.curveTo(xCurveValues[0], yCurveValues[0], xCurveValues[1], yCurveValues[1], xCurveValues[2],
            yCurveValues[2]);
        break;
      case 2:
        path.quadTo(xCurveValues[0], yCurveValues[0], xCurveValues[1], yCurveValues[1]);
        break;
      case 1:
        path.lineTo(xCurveValues[0], yCurveValues[0]);
        break;
    }
  }
```

Note the explanation comments, especially
**// For each given point (to which one specific x value in curveXValues belongs)
// use spline functions to get additional points for building the smooth path.
// But no more needed than maxPointsNeededForDrawing.**

**Comment by frigoref** [comment]
Was this "long paths" issue referring to my comment?

> Have you thought about the issue when the path line crosses a border to another territory which does not belong to the route?

**Comment by ron-murhammer** [comment]
@RoiEXLab I'd really like to see a few before and after screenshots of route drawing especially on common maps (standard ww2 maps, NWO, WaW, Napoleonic, 270BC, etc).

**Comment by RoiEXLab** [comment]
@ron-murhammer you may take a look at #634.
The bugs are caused by doubled route points on this map...
You may notice that the path is only smooth at the last two points

**Comment by ron-murhammer** [comment]
@RoiEXLab Yeah I saw that. Just want to make sure we don't fix it to look good on one map and then it looks much worse on most others.

**Comment by RoiEXLab** [comment]
@frigoref yes...

**Comment by RoiEXLab** [comment]
@frigoref would you mind forking my fork and creating a PR? This way we could do some progress on this PR before Friday evening GMT+1 Summer Time

**Comment by frigoref** [comment]
I'l try, but I am very busy until the weekend, so it might not work out in time.

**Comment by DanVanAtta** [comment]
Forking a fork is not necessary. Forks are simply remote branches where you have write access. The fork does not grant you write access to someone else fork. Branches exist as parallel entities, not parent-child.  There is no relationship between forks other than that they are different branches of the same project, same thing for the master branch or any other branch.

But, what you can do is (sorry, no clue how to do this without command line - git for windows has a right click 'bash from here', where you could copy paste these commands and then use eclipse GUI for anything else):
```
git add remote RoiEXlab git@github.com:RoiEXLab/triplea.git
git fetch RoiEXlab
  ## next command will checkout out RoiEXLab's branch
git checkout RoiEXlab/<branch_name>

  ## rename the branch to something suitable
git branch -m <new_branch_name>
  ## now the branch is renamed, push it to your remote repo. 
  ## note, git might want you to run a '--set-upstream-to=' command first, just follow the instructions it give
git push origin <new_branch_name>
# now use Github Web UI to submit a PR from <new_branch_name> into RoiEXlab/<branch_name>
```

**Comment by RoiEXLab** [comment]
@DanVanAtta I knew that... It just would be easier to do it like this...
As far as I know there is no option for github desktop, but I think when using git kraken there's a option for this...

**Comment by frigoref** [comment]
All possible on GitHub: I created a new branch on a common commit state between RoiEXLab RouteDrawer branch and my frigoref master branch. Afterwards did a PR from RoiEXLab RouteDrawer branch to my frigoref master branch: https://github.com/frigoref/triplea/tree/RoiEXLab_RouteDrawer

Now I can commit to this copy branch and create a PR for RoiEXLab branch. @RoiEXLab Could you help me focusing on what I am supposed commit? Should I put my coding above into a PR?

**Comment by RoiEXLab** [comment]
Just add your code which you suggested above instead of my line drawing etc.

**Comment by frigoref** [comment]
Here you go: https://github.com/RoiEXLab/triplea/pull/1

**Comment by RoiEXLab** [comment]
Merged... Feel free to add Screenshots...

**Comment by RoiEXLab** [comment]
Good Job @frigoref this is far better than the last try...
Your attempt to draw the route is more performant than my attempt.
Nonetheless it is still not perfect:
Not Smooth1
Not Smooth2

You may notice, that Beginning with the 3rd point back the path stops being smooth...

I edited the code to draw a straight line if the distance between start and end is not sufficient - therefore moved your code to a place where it is only called when the distance is sufficient...
This enabled me to remove your segment method checking how big the distance was...

Any ideas on how to get it smoother?

**Comment by frigoref** [comment]
Actually I am not seeing how what has changed :D
But from the commits I would argue that the straight line in case only 2 nodes exist is also coming through the method code getSmoothPath().
In any case, the second commit seems wrong as it assumes that there is always a number of 3 points available to draw a curve to the next target point of the path, but how do you ensure that? 

A possibility to continue smoothing would be to add path points from the interpolated values in case there is a long distance between two territory center path points. That way the long distance could be smoother further. However, that comes with the drawback of complexity and probably low additional value as in this case the path segment is more like a line than it is like a curve.

Currently I am more thinking how to employ the polygon information to ensure the path is crossing to the next territory only at appropriate borders. There is a funnel algorithm we could use, but I think this might be to much overhead and we can find a smarter way. 
One thought of mine was to buffer boarders between a territory and it's neighbors and enforce a path point on them.
You can see the issue I am addressing with this for example when clicking on the island Caralis and create a path to a sea zone 33.

**Comment by RoiEXLab** [comment]
>  In any case, the second commit seems wrong as it assumes that there is always a number of 3 points available to draw a curve to the next target point of the path, but how do you ensure that? 

I moved the method call inside the else brackets behind the if `numTerritoiries == 1 || points.length <= 2` check... Shouldn't this ensure that?

@frigoref Since you are really helping me out here, i decided to create another Fork on my organisation Account giving you write access to it... Check it out [here](https://github.com/RoiEXorg/triplea). If you agree, I will close this PR reopening it with the "new" fork

**Comment by frigoref** [comment]
Okay for the fork.
 
The other issue is not solved back the moving. Inside the getSmoothPath method for each points array entry a call must be made to the path object to add it. There are 3 ways available: a line, a quadratic curve and a b. curve. The difference between them is how many additional points are taken in to account for smoothing. Those additional points need to come from the interpolated points, but if the distance between the last added path point and the next is to short there won'the be enough interpolated points. Hence, the coding would fail.

**Comment by RoiEXLab** [comment]
This will be taken over by @frigoref

**Pull request closed without merging**